expo/expo-cli
# [config] Create dynamic config "app.config.js"

## Why 

Problems with the existing solution: `app.json`
- can't swap between modes like development and production
- no way to comment and document your config
- can't share parts of your config across multiple projects
- hard to discover what features are available 
- hard to know what everything in the config does
- config types aren't in the monorepo and can't be easily changed

## How

If a project has an `app.config.js`, that will take priority over an `app.json`. You can use the `app.config.js` to modify your `app.json` using environment variables or custom JS plugins. Here is how you might modify your project's `app.json` using the new `app.config.js`:
```js
// config is generated from the package.json and the app.json.
module.exports = function({ projectRoot, config, mode }) {
  // Use the mode to control how the app works in dev vs prod
  config.name = mode === 'development' ? 'dev app' : 'prod app';
  // Return the modified config
  return config;
}
```

Also add support for the following config types:
- `app.config.js`
- `app.config.json`
- ~~`app.config.json5`~~ - disabled via @ide request
- Added but disabled: 
  - ~~`app.config.ts`~~
  - ~~`app.config.yml`~~
  - ~~`app.config.yaml`~~
  - ~~`app.config.toml`~~

- I'm somewhat against the idea of having multiple different languages for the config. At the very least we could just add them later after `app.config.js` has migrated.
- Only allow sync configs to maintain usability with Next.js.

## Test Plan

- Add tests for reading from and validating each language
- Add a test for ensuring that `app.config` has higher priority than `app.json`
- Add a test to ensure the `app.config.js` method is passed the `app.json` and a `mode` option
- Ensure that other language `app` files don't get read (like `app.yml`, `app.json5`)
- Enforce that the config is always serialized to JSON so it can be used for the manifest.
